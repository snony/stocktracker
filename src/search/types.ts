export type Company = {name: string, symbol:string};
export type ClickResult = (company:Company) => void;

export type Companies = Array<Company>

export interface SearchProps{
    readonly getInfo: ClickResult;
    readonly companySymbols: Companies; 
}

export interface SearchState{
    readonly searchValue:string;
    readonly suggestedCompanies: Companies;
}