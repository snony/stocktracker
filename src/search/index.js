import SearchContainer from './searchContainer'
import { onSymbolSelected } from './searchContainer'
export default SearchContainer
export { onSymbolSelected }
