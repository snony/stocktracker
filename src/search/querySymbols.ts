import Fuse from 'fuse.js'

import { Company } from 'types'

const options = {
  threshold: 0.6,
  location: 0,
  distance: 10,
  keys: ['symbol', 'name']
}

const QuerySymbols = (queryString: string, db: Company[]) => {
  if (queryString === '') {
    return []
  }
  const fuse = new Fuse(db, options)
  const companyByNameOrSymbol = fuse.search(queryString)
  return companyByNameOrSymbol.length > 10
    ? companyByNameOrSymbol.slice(0, 10)
    : companyByNameOrSymbol
}

export default QuerySymbols
