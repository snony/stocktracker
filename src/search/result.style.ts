import { stockTrackerPadding } from 'index.styles'
import styled from 'styled'

export const Ul = styled('ul')`
  list-style-type: none;
  padding-left: 0;
  z-index: 1;
  position: absolute;
  width: calc(100% - 2 * ${stockTrackerPadding});
  background: ${props => props.theme.bgSearch};
`

export const Li = styled('li')`
  outline: none;
  padding: 1.5rem 1rem;
  color: white;
  &:hover {
    color: ${props => props.theme.linkHoverColour};
  }
  transition: 0.2s ease;
  font-size: 1rem;
`
