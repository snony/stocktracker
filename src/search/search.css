.search-container__input-wrapper {
  display: -ms-flexbox; /* IE10 */
  display: flex;
  align-items: center;
  width: 100%;
  padding: 0.7rem 0;
  border-bottom: var(--heading);
}

.search-container__icon {
  padding: 0.4rem 1rem 0.4rem 0;
  color: var(--primary-colour);
}

.search-container__input {
  height: 100%;
  width: 100%;
  background: transparent;
  border: transparent;
  color: white;
  font-size: 1.9rem;
  font-weight: 100;
  outline: transparent;
}

.search-container__results-list {
  list-style-type: none;
  padding-left: 0;
  z-index: 1;
  position: absolute;
  width: calc(100% - 2 * var(--stock-tracker-padding));
  background: var(--bg-search);
}

.search-container__result {
  color: var(--secondary-colour);
  outline: none;
  padding: 1.5rem 1rem;
  color: white;
  transition: 0.2s ease;
}

.search-container__result:hover {
  color: var(--link-visited);
}
