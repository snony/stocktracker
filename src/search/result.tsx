import './search.css'

import * as React from 'react'

import { Company } from '../types'

type CompanyOnSelect = (company: Company) => void
interface SearchResultProps {
  readonly results: Company[]
  readonly onClickResult: CompanyOnSelect
}

interface ResultProps {
  readonly company: Company
  readonly onClickResult: CompanyOnSelect
}

const Result: React.SFC<ResultProps> = ({ company, onClickResult }) => {
  const click = () => onClickResult(company)

  return (
    <li onClick={click} className="search-container__result">
      {company.name} ({company.symbol})
    </li>
  )
}

class SearchResults extends React.PureComponent<SearchResultProps> {
  public render() {
    return (
      <ul className="search-container__results-list">
        {this.props.results.map(company => (
          <Result key={company.symbol} company={company} onClickResult={this.props.onClickResult} />
        ))}
      </ul>
    )
  }
}
export default SearchResults
