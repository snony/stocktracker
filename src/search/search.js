import React from 'react'

import SearchResults from './result'
import QueryDB from './queryDB'

import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'

import './search.css'

class Search extends React.Component {
  constructor(props) {
    super(props)
    this.state = {
      value: '',
      suggestedCompanies: []
    }
  }

  handleInputChange = ({ target: { value } }) => {
    const suggestedCompanies = !!value ? QueryDB(value, this.props.companiesDB) : []
    this.setState({ value, suggestedCompanies })
  }

  onClickResult = company => {
    const searchValue = company.name + ' ' + company.symbol
    this.props.getInfo(company)
    this.setState({ value: searchValue, suggestedCompanies: [] })
  }

  render() {
    const suggestedCompanies = this.state.suggestedCompanies
    const searchValue = this.state.value
    return (
      <div className="search-container">
        <div className="search-container__input-wrapper">
          <span className="search-container__icon">
            <FontAwesomeIcon icon="search" size="lg" />
          </span>
          <input
            type="text"
            className="search-container__input"
            value={searchValue}
            onChange={this.handleInputChange}
          />
        </div>
        <SearchResults results={suggestedCompanies} onClickResult={this.onClickResult} />
      </div>
    )
  }
}

export default Search
