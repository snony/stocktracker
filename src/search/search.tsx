import React from 'react'

import SearchResults from './result'
import QuerySymbols from './querySymbols'

import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'

import './search.css'
import { Company, SearchProps, SearchState } from './types';



const defaultValue = 'Apple Inc. (AAPL)'
const suggestedCompanies: Company[] = []
const initState = {searchValue:defaultValue, suggestedCompanies}

class Search extends React.PureComponent<SearchProps, SearchState> {
  readonly state : SearchState = initState
  
  private handleInputChange = ({ target: { value } }:{target:{value:string}}) => {
    const suggestedCompanies = !!value ? QuerySymbols(value, this.props.companySymbols) : []
    this.setState({ searchValue:value, suggestedCompanies })
  }

  private onClickResult = (company:Company) => {
    const searchValue = `${company.name} (${company.symbol})`
    this.props.getInfo(company)
    this.setState({ searchValue, suggestedCompanies: [] })
  }

  render() {
    return (
      <div className="search-container">
        <div className="search-container__input-wrapper">
          <span className="search-container__icon">
            <FontAwesomeIcon icon="search" size="lg" />
          </span>
          <input
            type="text"
            className="search-container__input"
            value={this.state.searchValue}
            onChange={this.handleInputChange}
          />
        </div>
        <SearchResults results={this.state.suggestedCompanies} onClickResult={this.onClickResult} />
      </div>
    )
  }
}

export default Search
