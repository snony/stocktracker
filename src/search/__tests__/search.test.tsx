import { getCompanySymbols } from '__mock__/companySymbols.mock'
import { mount, shallow } from 'enzyme'
import FetchStatus from 'fetchStatus';
import React from 'react'
import renderer from 'react-test-renderer'
import Search from 'search/search'
import { SearchProps } from 'search/types'

describe('Search Component', () => {
  it('should render correctly when no company is searched and fetchStatus is initial', () => {
    const searchProps = { fetchStatus: FetchStatus.INITIAL } as SearchProps
    const tree = renderer.create(<Search {...searchProps} />).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('should render correctly when there is a search but company symbols has a failure fetch status ', () => {
    const searchProps = { fetchStatus: FetchStatus.FAILED } as SearchProps
    const tree = renderer.create(<Search {...searchProps} />).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('should render correctly with search results', () => {
    const searchProps = {} as SearchProps
    const tree = mount(<Search {...searchProps} />)
    tree.setState({
      companies: getCompanySymbols(2)
    })
    expect(tree).toMatchSnapshot()
  })

  it('should handle search input', () => {
    const searchProps = { companySymbols: [{ name: 'Apple Inc', symbol: 'Aapl' }] } as SearchProps
    const tree = shallow(<Search {...searchProps} />)
    const newSearchValue = { target: { value: 'Facebook' } }
    tree.find('input').simulate('change', newSearchValue)
    expect(tree.state()).toEqual({ searchValue: newSearchValue.target.value, companies: [] })
  })

  it('should handle Company on Select', () => {
    const searchProps: SearchProps = { companySymbols: [], getInfo: jest.fn(), fetchStatus: FetchStatus.SUCCESS }
    const tree = mount(<Search {...searchProps} />)
    tree.setState({ companies: [{ name: 'Facebook', symbol: 'fb' }] })
    tree.find('li').simulate('click')
    expect(tree.state()).toEqual({ searchValue: 'Facebook (fb)', companies: [] })
  })
})
