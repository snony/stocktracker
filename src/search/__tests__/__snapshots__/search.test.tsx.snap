// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search Component should render correctly when no company is searched and fetchStatus is initial 1`] = `
.emotion-1 {
  height: 100%;
  width: 100%;
  background: transparent;
  border: transparent;
  color: white;
  font-size: 1.9rem;
  font-weight: 100;
  outline: transparent;
}

.emotion-3 {
  display: -ms-flexbox;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: 100%;
  padding: 0.7rem 0;
}

.emotion-0 {
  padding: 0.4rem 1rem 0.4rem 0;
}

.emotion-2 {
  color: white;
  display: block;
  margin-bottom: 0.3rem;
  font-size: 0.9rem;
}

<div>
  <div
    className="emotion-3"
  >
    <span
      className="emotion-0"
    >
      <svg
        aria-hidden="true"
        className="svg-inline--fa fa-search fa-w-16 fa-lg "
        data-icon="search"
        data-prefix="fas"
        role="img"
        style={Object {}}
        viewBox="0 0 512 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M505 442.7L405.3 343c-4.5-4.5-10.6-7-17-7H372c27.6-35.3 44-79.7 44-128C416 93.1 322.9 0 208 0S0 93.1 0 208s93.1 208 208 208c48.3 0 92.7-16.4 128-44v16.3c0 6.4 2.5 12.5 7 17l99.7 99.7c9.4 9.4 24.6 9.4 33.9 0l28.3-28.3c9.4-9.4 9.4-24.6.1-34zM208 336c-70.7 0-128-57.2-128-128 0-70.7 57.2-128 128-128 70.7 0 128 57.2 128 128 0 70.7-57.2 128-128 128z"
          fill="currentColor"
          style={Object {}}
        />
      </svg>
    </span>
    <input
      className="emotion-1"
      onChange={[Function]}
      type="text"
      value="Apple Inc. (AAPL)"
    />
    <span
      className="emotion-2"
    >
      ...Loading Suggested companies
    </span>
  </div>
</div>
`;

exports[`Search Component should render correctly when there is a search but company symbols has a failure fetch status  1`] = `
.emotion-1 {
  height: 100%;
  width: 100%;
  background: transparent;
  border: transparent;
  color: white;
  font-size: 1.9rem;
  font-weight: 100;
  outline: transparent;
}

.emotion-3 {
  display: -ms-flexbox;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: 100%;
  padding: 0.7rem 0;
}

.emotion-0 {
  padding: 0.4rem 1rem 0.4rem 0;
}

.emotion-2 {
  color: white;
  display: block;
  margin-bottom: 0.3rem;
  font-size: 0.9rem;
}

<div>
  <div
    className="emotion-3"
  >
    <span
      className="emotion-0"
    >
      <svg
        aria-hidden="true"
        className="svg-inline--fa fa-search fa-w-16 fa-lg "
        data-icon="search"
        data-prefix="fas"
        role="img"
        style={Object {}}
        viewBox="0 0 512 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M505 442.7L405.3 343c-4.5-4.5-10.6-7-17-7H372c27.6-35.3 44-79.7 44-128C416 93.1 322.9 0 208 0S0 93.1 0 208s93.1 208 208 208c48.3 0 92.7-16.4 128-44v16.3c0 6.4 2.5 12.5 7 17l99.7 99.7c9.4 9.4 24.6 9.4 33.9 0l28.3-28.3c9.4-9.4 9.4-24.6.1-34zM208 336c-70.7 0-128-57.2-128-128 0-70.7 57.2-128 128-128 70.7 0 128 57.2 128 128 0 70.7-57.2 128-128 128z"
          fill="currentColor"
          style={Object {}}
        />
      </svg>
    </span>
    <input
      className="emotion-1"
      onChange={[Function]}
      type="text"
      value="Apple Inc. (AAPL)"
    />
    <span
      className="emotion-2"
    >
      Cannot search Company: check connection
    </span>
  </div>
</div>
`;

exports[`Search Component should render correctly with search results 1`] = `
.emotion-1 {
  height: 100%;
  width: 100%;
  background: transparent;
  border: transparent;
  color: white;
  font-size: 1.9rem;
  font-weight: 100;
  outline: transparent;
}

.emotion-2 {
  display: -ms-flexbox;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: 100%;
  padding: 0.7rem 0;
}

.emotion-0 {
  padding: 0.4rem 1rem 0.4rem 0;
}

<Search>
  <div>
    <Styled(div)>
      <div
        className="emotion-2"
      >
        <Styled(span)>
          <span
            className="emotion-0"
          >
            <FontAwesomeIcon
              border={false}
              className=""
              fixedWidth={false}
              flip={null}
              icon="search"
              inverse={false}
              listItem={false}
              mask={null}
              pull={null}
              pulse={false}
              rotation={null}
              size="lg"
              spin={false}
              symbol={false}
              title=""
              transform={null}
            >
              <svg
                aria-hidden="true"
                className="svg-inline--fa fa-search fa-w-16 fa-lg "
                data-icon="search"
                data-prefix="fas"
                role="img"
                style={Object {}}
                viewBox="0 0 512 512"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M505 442.7L405.3 343c-4.5-4.5-10.6-7-17-7H372c27.6-35.3 44-79.7 44-128C416 93.1 322.9 0 208 0S0 93.1 0 208s93.1 208 208 208c48.3 0 92.7-16.4 128-44v16.3c0 6.4 2.5 12.5 7 17l99.7 99.7c9.4 9.4 24.6 9.4 33.9 0l28.3-28.3c9.4-9.4 9.4-24.6.1-34zM208 336c-70.7 0-128-57.2-128-128 0-70.7 57.2-128 128-128 70.7 0 128 57.2 128 128 0 70.7-57.2 128-128 128z"
                  fill="currentColor"
                  style={Object {}}
                />
              </svg>
            </FontAwesomeIcon>
          </span>
        </Styled(span)>
        <input
          className="emotion-1"
          onChange={[Function]}
          type="text"
          value="Apple Inc. (AAPL)"
        />
      </div>
    </Styled(div)>
  </div>
</Search>
`;
